### ISSUE TITLE
Preprocessing External Images from S3 

### ISSUE THREAD

**Issue by LpmRaven** [gatsbyjs/gatsby#22746]
@lannonbr, I'm hoping you can help. I have a question regarding the documentation you wrote for '[Preprocessing External Images](https://www.gatsbyjs.org/docs/preprocessing-external-images/)' #17385 , it's quite a deviation from the example that you provided so ill provide some context.

I am building an e-commerce store, which contains products (obv), each of the product pages is created programmatically from gatsby-node. I have images for each of these products in an S3 bucket. The data for each product contains `image1`, which is the name (key) of the file in the S3 bucket. 
```graphql
const productQuery = `
query ($nextToken: String, $region: AWSAppSync_Region!, $productType: String!) {
    appSync {
        productsForCategory(
        region: $region,
        productTypeCategorySlugSubCategorySlugRatingsTotalSavingsVsRrpValue: {
            beginsWith: {
            productType: $productType
            }
        },
        nextToken: $nextToken
        ) {
        items {
            globalIdColor
            productType
            region
            categorySlug
            subCategorySlug
            image1
        }
        nextToken
        }
    }
}`;
```
I wish to attach the image as a File node to this product item in the GraphQL layer.

Currently, I request my images in what feels like a really wrong way. I gain access to my images using the `gatsby-source-s3-image` plugin. This allows me to access the images through allImageSharp. But then I need to match each product to an image. On each product page, I have an image component which is passed the 'image1' as `props.filename`.

```javascript
export const Image = props => {

  return (
    <StaticQuery
      query={graphql`
      query {
        images: allImageSharp {
          edges {
            node {
                square: fluid(maxWidth: 300, quality: 100) {
                  ...GatsbyImageSharpFluid
                  originalName
                }
                thumbnailSizes: fluid(maxWidth: 256, quality: 100) {
                  ...GatsbyImageSharpFluid
                  originalName
                }
                largeSizes: fluid(maxWidth: 400, quality: 100) {
                  ...GatsbyImageSharpFluid
                  originalName
                }
            }
          }
        }
      } 
    `}
      render={data => {
        const image = data && data.images && data.images.edges && data.images.edges.find(n => {
          if (n.node.square.originalName && props.filename) {
            const key = n.node.square.originalName.toLowerCase();
            const filename = props.filename.toLowerCase();

            return key.includes(filename);
          }

          return false;
        });

        if (!image) {
          return (
            <ImageUnavailable />
          );
        }

        if (image) {
          return <Img  fluid={image.node.largeSizes} />;
        }
        if (props.thumbnail && image) {
          return <Img  fluid={image.node.thumbnailSizes} />;
        }
        if (props.blogPost && image) {
          return <Img  fluid={image.node.square} />;
        }
      }}
    />
  )
};
```
This only actually works 50% of the time even if the product key is in the S3 bucket, maybe something to do with this misuse of static query or that there is thousands of images.
I'm hoping that if I can attach the image as a File node inside gatsby-node, then I can delete this 'find and includes' way of getting my product image. 

So my question, how do I attach the image as a File node for each product item from a graphql request rather than on the creation of a MarkdownRemark type?

**Comment by Js-Brecht** [comment]
Hey @LpmRaven, are you loading the data retrieved by `productQuery` into nodes in Gatsby's GraphQL layer?

**Comment by LpmRaven** [comment]
Hi @Js-Brecht, the `productQuery` is inside gatsby-node.js and is used to programmatically generate all of the product pages. I am trying to attach the image in S3 as a file node to this layer, so I can query it within the`productQuery`. I believe this data in `productQuery` already resides in nodes in gatsby GraphQL layer as I retrieve it from AWS Appsync using the `gatsby-source-graphql` plugin. I may be misunderstanding you.

**Comment by Js-Brecht** [comment]
> I believe this data in `productQuery` already resides in nodes in gatsby GraphQL layer as I retrieve it from AWS Appsync using the `gatsby-source-graphql` plugin.

Ahh, I see.  The reason I was asking is because if you were creating the nodes yourself, you could potentially attach the image `File` node as you go.  But the fact that it's a 3rd party schema changes things.

I'm really not very familiar with the use of 3rd party schemas, so I may have to generalize a bit.  You should be able to use a [custom resolver](https://www.gatsbyjs.org/docs/node-apis/#createResolvers) to do what you're doing.  If you can identify the root Node typename of your items (`AppSync_Items`, maybe?), then you can define a field on it.  Inside of the resolver for that field, you can do a query with `context.nodeModel` for `File` types, and filter it by `base: { eq: source.image1 }`.  Then you return that resolved node's `childImageSharp`.

It may look something like this, but YMMV:
```js
exports.createResolvers = ({ createResolvers }) => {
    const resolvers = {
        AppSync_Items: {
            image1: {
                type: 'ImageSharp',
                resolve: async (source, args, context, info) => {
                    const node = await context.nodeModel.runQuery({
                        query: {
                            filter: {
                                base: { eq: source.image1 }
                            }
                        },
                        type: 'File'
                    });
                    return node.childImageSharp;
                }
            },
        }
    }
    createResolvers(resolvers)
}
```

Now, in your page query, you would just use that `image` field as if it is a `childImageSharp`.

```graphql
query ItemQuery($itemId: String!) {
    item(id: $itemId) {
        image1 {
            square: fluid(maxWidth: 300, quality: 100) {
                ...GatsbyImageSharpFluid
            }
            thumbnailSizes: fluid(maxWidth: 256, quality: 100) {
                ...GatsbyImageSharpFluid
            }
            largeSizes: fluid(maxWidth: 400, quality: 100) {
                ...GatsbyImageSharpFluid
            }
        }
    }
}
```
It's hard for me to know how to design this specifically for your use, because I don't know what your schema looks like (or what your gatsby-node/page queries look like, for that matter! :laughing:).  You can examine the schema yourself, though, by using `gatsby repl`, and executing the [schema` command](https://www.gatsbyjs.org/docs/gatsby-repl/#schema).  Perhaps you can adapt this to work for you.

---


A more generic approach would be to add a new root node that will act as a query entry point to find the image file for you.  Something sort of like this:
```js
exports.createResolvers = ({ createResolvers }) => {
    const resolvers = {
        Query: {
            productImage: {
                type: 'ImageSharp',
                args: {
                    imageName: 'String!'
                },
                resolve: async (source, args, context, info) => {
                    const node = await context.nodeModel.runQuery({
                        query: {
                            filter: {
                                base: args.imageName
                            }
                        },
                        type: 'File'
                    });
                    return node.childImageSharp;
                }
            },
        }
    }
    createResolvers(resolvers)
}
```
Make sure you pass `image1` as a context parameter to each page, so you can query it like this:
```graphql
query ItemQuery($image1: String!) {
    productImage(imageName: $image1) {
        square: fluid(maxWidth: 300, quality: 100) {
            ...GatsbyImageSharpFluid
        }
        thumbnailSizes: fluid(maxWidth: 256, quality: 100) {
            ...GatsbyImageSharpFluid
        }
        largeSizes: fluid(maxWidth: 400, quality: 100) {
            ...GatsbyImageSharpFluid
        }
    }
}
```

---

You could even take this a step further, if you really wanted to, by downloading only the images you need from your S3 bucket, as opposed to downloading **ALL** of them, and then querying for them.  That's a bit out of scope here, though.

**Comment by LpmRaven** [comment]
@Js-Brecht I see. From your suggestion, I took a look at my schema and this createResolvers (below) adds a new imageFile field to the item, but I don't really understand how to use context.nodeModel.runQuery even after reading the docs.

```
exports.createResolvers = ({ createResolvers }) => {
    const resolvers = {
        AWSAppSync_Product: {
            imageFile: {
                type: 'ImageSharp',
                resolve: async (source, args, context, info) => {
                    const node = await context.nodeModel.runQuery({
                        query: {
                            filter: {
                                ??
                            }
                        },
                        type: 'File'
                    });
                    return node.childImageSharp;
                }
            },
        }

    }
    createResolvers(resolvers)
}
```
The data is on a different node (s3ImageAsset), if I was to query the image from S3 I would use the query below, so how do I attach this to imageFile on AWSAppSync_Product? Will the query filter for File types `base: { eq: source.image1 }` look for this? 
```
query MyS3Query {
  s3ImageAsset(Key: {eq: ""}) {
    Key
    absolutePath
    childImageSharp {
      fluid
    }
  }
}
```

**Comment by Js-Brecht** [comment]
You can have it query for that type, too.  Just change `type: 'File'` to be `type: 'S3ImageAsset'`.

The reason I mentioned querying for a `File` node is because that plugin uses [createRemoteFileNode()](https://github.com/jessestuart/gatsby-source-s3-image/blob/master/src/source-nodes.ts#L101-L108) to download the files from the image URLs, which creates `File` nodes.  The `base` property on `File` nodes will usually be the same as the `originalName` on the `ImageSharp` node, so it was just a most direct route to access it.

On the other hand, I think that `base` and `originalName` should be the same as the `Key` property on the `S3ImageAsset` type, based on [how the URL is constructed](https://github.com/jessestuart/gatsby-source-s3-image/blob/master/src/utils.ts#L97-L99).  I can't say for sure without seeing the data.  But if that data is what matches your `image1` field, then you can use it just as well.  Something like this, then:
```js
exports.createResolvers = ({ createResolvers }) => {
    const resolvers = {
        AWSAppSync_Product: {
            imageFile: {
                type: 'ImageSharp',
                resolve: async (source, args, context, info) => {
                    const node = await context.nodeModel.runQuery({
                        query: {
                            filter: {
                                Key: { eq: source.image1 }
                            }
                        },
                        type: 'S3ImageAsset' // <----- updated
                    });
                    return node.childImageSharp;
                }
            },
        }

    }
    createResolvers(resolvers)
}
```

**Comment by Js-Brecht** [comment]
[Here is the docs](https://www.gatsbyjs.org/docs/node-model/#runQuery) for `runQuery`.  I think I may have forgotten an argument: `firstOnly`.  That will be important, otherwise you will get an array.

**Comment by LpmRaven** [comment]
Thanks for explaining that, it makes much more sense now. 

The problem I was failing to debug is that the node response does not contain childImageSharp, which is why it is returning null. I'm not sure why this is.
When I query the data through localhost GraphQL playground, the childImageSharp is not missing.

Sidenote: With `firstOnly` its actually just returning a single object so no need for the array length check.

## Playground:

```graphql
query MyQuery {
  file(base: {eq: "FAJKRAM20001-blue.jpg"}) {
    childImageSharp {
      id
      fluid {
        src
      }
    }
  }
}
```

```json
{
  "data": {
    "file": {
      "childImageSharp": {
        "id": "a10ee426-9054-5e21-81a0-c0f49b454a8b",
        "fluid": {
          "src": "/static/bd6f771a35ccc879a57b61b473d95fd3/14b42/FAJKRAM20001-blue.jpg"
        }
      }
    }
  }
}
```
## gatsby-node.js
```js
exports.createResolvers = ({ createResolvers }) => {
    const resolvers = {
        AWSAppSync_Product: {
            imageFile: {
                type: 'ImageSharp',
                resolve: async (source, args, context, info) => {
                    console.log('source.image1', source.image1);

                    const node = await context.nodeModel.runQuery({
                        query: {
                            filter: {
                                base: { eq: source.image1 }
                            }
                        },
                        type: 'File',
                        firstOnly: true
                    });
                    console.log(node)

                    return node.childImageSharp;
                }
            },
        }

    }
    createResolvers(resolvers)
}
```

```
{
  id: '0661af13-f20e-5094-9454-dd2867ea1f0e',
  children: [ 'a10ee426-9054-5e21-81a0-c0f49b454a8b' ],
  parent: null,
  internal: {
    contentDigest: 'bd6f771a35ccc879a57b61b473d95fd3',
    type: 'File',
    mediaType: 'image/jpeg',
    description: 'File "<redacted>/FAJKRAM20001-blue.jpg"',
    counter: 403,
    owner: 'gatsby-source-filesystem'
  },
  sourceInstanceName: '__PROGRAMMATIC__',
  absolutePath: '/Users/name/Workspace/w-gatsby/.cache/gatsby-source-filesystem/281861219ae7abbf1535779ca443afe4/FAJKRAM20001-blue.jpg',
  relativePath: '.cache/gatsby-source-filesystem/281861219ae7abbf1535779ca443afe4/FAJKRAM20001-blue.jpg',
  extension: 'jpg',
  size: 188412,
  prettySize: '188 kB',
  modifiedTime: '2020-04-05T08:28:42.653Z',
  accessTime: '2020-04-05T08:28:41.322Z',
  changeTime: '2020-04-05T08:28:42.655Z',
  birthTime: '2020-04-05T08:28:41.322Z',
  root: '/',
  dir: '/Users/name/Workspace/w-gatsby/.cache/gatsby-source-filesystem/281861219ae7abbf1535779ca443afe4',
  base: 'FAJKRAM20001-blue.jpg',
  ext: '.jpg',
  name: 'FAJKRAM20001-blue',
  relativeDirectory: '.cache/gatsby-source-filesystem/281861219ae7abbf1535779ca443afe4',
  dev: 16777220,
  mode: 33188,
  nlink: 1,
  uid: 501,
  gid: 20,
  rdev: 0,
  blksize: 4096,
  ino: 14456866,
  blocks: 392,
  atimeMs: 1586075321321.645,
  mtimeMs: 1586075322652.821,
  ctimeMs: 1586075322654.8367,
  birthtimeMs: 1586075321321.645,
  atime: '2020-04-05T08:28:41.322Z',
  mtime: '2020-04-05T08:28:42.653Z',
  ctime: '2020-04-05T08:28:42.655Z',
  birthtime: '2020-04-05T08:28:41.322Z',
  url: '<redacted>'
}
```

**Comment by Js-Brecht** [comment]
Oh!  I’m sorry.  This is querying the nodes directly, so it doesn’t have the “foreign key” relationship.  It’s not going through the GraphQL resolver.

You can either do another query for `type: ‘ImageSharp’` with the `id ` field equal to that first element in `children`, or you can find the field on type `ImageSharp` that matches the value you need, and filter on that in the original query instead.

**Comment by Js-Brecht** [comment]
I was wrong about `getNode`.  It’s [`getNodeById`]( 
https://www.gatsbyjs.org/docs/node-model/#getNodeById) 😂.

**Comment by Js-Brecht** [comment]
Just had a sec to come back to this (after getting some sleep).  You don't have to bother returning `childImageSharp` from the resolver, or even resolving the `ImageSharp` node via `getNodeById`.  All you have to do is return the `File` node as is, and in GraphQL, you can query the `childImageSharp` field.

Like this:
```js
exports.createResolvers = ({ createResolvers }) => {
    const resolvers = {
        AWSAppSync_Product: {
            imageFile: {
                type: 'File', // this needs to be 'File', since it's not returning an 'ImageSharp' node anymore
                resolve: (source, args, context, info) => {
                    // A promise is expected to be returned anyway so don't need
                    // to `await` the result here, if all we're doing is returning
                    // the `File` node
                    return context.nodeModel.runQuery({
                        query: {
                            filter: {
                                base: { eq: source.image1 }
                            }
                        },
                        type: 'File',
                        firstOnly: true
                    });
                }
            },
        }
    }
    createResolvers(resolvers)
}
```
Then, in general, query like this:
```graphql
query {
    type {
        imageFile {
            childImageSharp {
                fluid {
                    ...GatsbyImageSharpFluid
                }
            }
        }
    }
}
```

GraphQL will handle resolving the `ImageSharp` node for you.  Simpler that way, and simpler is better, right?

**Comment by LpmRaven** [comment]
Great, thanks @Js-Brecht! Got that all working, I can't thank you enough.

**Comment by rwchampin** [comment]
Hey guys!

Im facing the EXACT same issue.  I was wondering how your graphql.schema looked like?  Everytime I try to do this, it tells me that the type doesnt exist on the schema.  Im using amplify and gatsby so I have an RDL schema.  Like...where are you defining your model type of AWSAppSync_Product from?
